internal/datanode/compaction/segment_writer.go

```diff
@@ -0,0 +1,165 @@
+// SegmentInsertBuffer can be reused to buffer all insert data of one segment
+// buffer.Serialize will serialize the InsertBuffer and clear it
+// pkstats keeps tracking pkstats of the segment until Finish
+
+package compaction
+
+import (
+	"fmt"
+	"math"
+
+	"go.uber.org/atomic"
+
+	"github.com/milvus-io/milvus-proto/go-api/v2/schemapb"
+	"github.com/milvus-io/milvus/internal/datanode/writebuffer"
+	"github.com/milvus-io/milvus/internal/proto/etcdpb"
+	"github.com/milvus-io/milvus/internal/storage"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/paramtable"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+)
+
+type SegmentWriter struct {
+	writer  *storage.SerializeWriter[*storage.Value]
+	closers []func() (*storage.Blob, error)
+	tsFrom  typeutil.Timestamp
+	tsTo    typeutil.Timestamp
+
+	pkstats      *storage.PrimaryKeyStats
+	segmentID    int64
+	partitionID  int64
+	collectionID int64
+	sch          *schemapb.CollectionSchema
+	rowCount     *atomic.Int64
+}
+
+func (w *SegmentWriter) GetRowNum() int64 {
+	return w.rowCount.Load()
+}
+
+func (w *SegmentWriter) GetCollectionID() int64 {
+	return w.collectionID
+}
+
+func (w *SegmentWriter) GetPartitionID() int64 {
+	return w.partitionID
+}
+
+func (w *SegmentWriter) GetSegmentID() int64 {
+	return w.segmentID
+}
+
+func (w *SegmentWriter) GetPkID() int64 {
+	return w.pkstats.FieldID
+}
+
+func (w *SegmentWriter) Write(v *storage.Value) error {
+	ts := typeutil.Timestamp(v.Timestamp)
+	if ts < w.tsFrom {
+		w.tsFrom = ts
+	}
+	if ts > w.tsTo {
+		w.tsTo = ts
+	}
+
+	w.pkstats.Update(v.PK)
+	w.rowCount.Inc()
+	return w.writer.Write(v)
+}
+
+func (w *SegmentWriter) Finish(actualRowCount int64) (*storage.Blob, error) {
+	w.writer.Flush()
+	codec := storage.NewInsertCodecWithSchema(&etcdpb.CollectionMeta{ID: w.collectionID, Schema: w.sch})
+	return codec.SerializePkStats(w.pkstats, actualRowCount)
+}
+
+func (w *SegmentWriter) IsFull() bool {
+	w.writer.Flush()
```

Comment:
Why do we need to Flush each time we check IsFull?

Reply:
So donot check isFull too frequently, its expensive